oauth-common/src/main/java/io/strimzi/kafka/oauth/validator/OAuthIntrospectionValidator.java

```diff
@@ -210,7 +213,8 @@ private void performOptionalChecks(JsonNode response) {
 
         if (audience != null) {
             value = response.get("aud");
-            if (value == null || !audience.equals(value.asText())) {
+            List<String> audienceList = value != null ? JSONUtil.asListOfString(value) : Collections.emptyList();
```

Comment:
When `value` is a JSON string, rather than a JSON array of strings, `JSONUtil.asListOfString` will split it on whitespace, which doesn't seem to be correct according to https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-25 or https://openid.net/specs/openid-connect-core-1_0.html#IDToken. It _might_ be, because "The interpretation of audience values is generally application specific.". 

The JWT document also says that the content of the JSON strings can also be a URI if the String contains a `:`, Is that something we need to think about here?

Reply:
Good catch about `JSONUtil.asListOfString()` even though I think it behaves slightly different than you described :)

If the value is `"kafka"` then it's a text node, and it will be correctly parsed. 
If the value is `"kafka rest-api"` - that's the application specific handling, and not by spec - it will be correctly parsed under the assumption that the client id can't contain a space character (I have not checked this assumption). 
If the value is `"kafka, rest-api"` it is a text node again and not by the spec - it WON'T be correctly parsed unless there is client id containing a comma and a space (I assume these are not valid characters for client id). I also assume this case would never happen that multiple client ids would be sent this way - as a single string.
If the value is `["kafka", "rest-api"]` or `["kafka"]` it is an array node and will be handled correctly.

We could improve the JSONUtil.asListOfString() where it parses the list of strings sent as a single string to also treat the comma as a separator, not just the space. Assuming that all the contexts where it is used don't allow comma and space as legal characters inside individual string items or that no reasonable person would use these characters in client ID.

The spec indeed says that it could be a url, but I'm unaware of any implementation using that, and if we go testing proper support if url is the value then we're possibly spending time on something that no-one would ever use and that has no real implementation for us to test with.

Reply:
The spec seems to be silent about whether the string can contain whitespace or not. I assumed they could, but that such values should not be treated as a list of tokens, but rather a single value containing a string. In any case I'll defer to you on this, I just wanted to point this stuff out.

Reply:
Let's just leave it like this for now.